File: DataStructures-Algorithms/1. Arrays/Largest_sum_contiguous_subarray_2.c
/*
Write an efficient program to find the sum of contiguous subarray within a one-dimensional array of numbers which has the largest sum. 

Time Complexity : O(n^2)
Space Complexity : O(n)
*/

#include<stdio.h>
#include<limits.h>

void maxSubArraySum( int arr[], int n)
{
	int  sumArr[n], maxSum = INT_MIN;
	
	sumArr[0] = arr[0];
	
	for(int i=1; i<n; ++i)
		sumArr[i] = sumArr[i-1]+arr[i]; 
		
	for(int i=0; i<n; ++i)
		maxSum = sumArr[i] > maxSum ? sumArr[i] : maxSum; 
	
	for(int i=1; i<n; ++i)
	{
		int prevSum = sumArr[i-1];
		
		for(int j=i; j<n; ++j)
		{
			int tempSum = sumArr[j] - prevSum;
			
			maxSum = tempSum > maxSum ? tempSum : maxSum;
		}
	}

	printf(" Max. sum in a contiguous Subarray : %d ",maxSum);
}

void main()
{
	int arr[] = {-3, 10, 5, -20, 4, 11, -5, 7};
	
	int n = sizeof(arr)/sizeof(*arr);
	
	maxSubArraySum(arr,n);
}

File: CODING/12. Dynamic Programming/Largest_sum_contiguous_subarray_2.java
/* 
 * Write an efficient program to find the sum of contiguous subarray within a one-dimensional array of numbers that has the largest sum.
 *
 * The implementation handles the case when all numbers in array are negative.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 *  */

import java.util.*;

class Kadane {

	private int maxSubArraySum(int arr[]) {

		int size = arr.length;

		int max_sum_so_far = arr[0],
		curr_max = arr[0];

		for(int i = 1; i < size; ++i) {

			curr_max = Math.max(arr[i], curr_max + arr[i]);
			max_sum_so_far = Math.max(max_sum_so_far, curr_max);
		}

		return max_sum_so_far;
	}

	public static void main(String args[]) {

		int arr[] = { -2, -3, 4, -1, -2, 1, 5, -3};
		Kadane obj = new Kadane();
		System.out.println("Maximum contiguous sum is " + obj.maxSubArraySum(arr));
	}
}
